### PR TITLE
Add cleanup chain element

### PR DESCRIPTION
**Pull request by glazychev-art**
Signed-off-by: Artem Glazychev <artem.glazychev@xored.com>

## Description
This PR adds new chain element that cleans up resources on `chainCtx.Done()`

## Issue link
Issue: https://github.com/networkservicemesh/sdk-vpp/issues/516

## How Has This Been Tested?

- [x] Added unit testing to cover
- [ ] Tested manually
- [x] Tested by integration testing
- [ ] Have not tested

## Types of changes

- [ ] Bug fix
- [x] New functionallity
- [ ] Documentation
- [ ] Refactoring
- [ ] CI

### CODE REVIEW

File: pkg/networkservice/common/cleanup/client.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+			// Add to metadata if we want to delete clientconn
+			if c.ccClose {
+				storeCC(ctx)
+			}
+
+			<-factory.Close(begin.CancelContext(cancelCtx))
+			atomic.AddInt32(&c.activeConns, -1)
+
+			if atomic.LoadInt32(&c.activeConns) == 0 && c.doneCh != nil {
+				c.doneCh <- struct{}{}
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/cleanup/client.go:88]:
```suggestion
				close(c.doneCh)
```

File: pkg/networkservice/common/cleanup/client.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	if cancel, ok := loadAndDeleteCancel(ctx); ok {
+		cancel()
+	}
+	cancelCtx, cancel := context.WithCancel(context.Background())
+	storeCancel(ctx, cancel)
+
+	factory := begin.FromContext(ctx)
+	go func() {
+		// Update active connections counter. Needed for a cleanup done notification.
+		atomic.AddInt32(&c.activeConns, 1)
```
[the remark this answers is not in the document]
Comment (glazychev-art) [reply pkg/networkservice/common/cleanup/client.go]: Yes. Thank you. Done

File: pkg/networkservice/common/cleanup/client.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	chainCtx context.Context
+
+	ccClose     bool
+	doneCh      chan<- struct{}
+	activeConns int32
+}
+
+// NewClient - returns a cleanup client chain element
+func NewClient(ctx context.Context, opts ...Option) networkservice.NetworkServiceClient {
+	o := &option{}
```
[the remark this answers is not in the document]
Comment (glazychev-art) [reply pkg/networkservice/common/cleanup/client.go]: done

File: pkg/networkservice/common/cleanup/client.go (hunk 1 above)
Comment (glazychev-art) [reply pkg/networkservice/common/cleanup/client.go]: Done

File: pkg/networkservice/common/cleanup/client.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	}
+	c := &cleanupClient{
+		chainCtx: ctx,
+		ccClose:  o.ccClose,
+		doneCh:   o.doneCh,
+	}
+	go func() {
+		<-c.chainCtx.Done()
+		if atomic.LoadInt32(&c.activeConns) == 0 && c.doneCh != nil {
+			c.doneCh <- struct{}{}
```
[the remark this answers is not in the document]
Comment (glazychev-art) [reply pkg/networkservice/common/cleanup/client.go]: Done

**Review by Chunosov — APPROVED** [review]

File: pkg/networkservice/common/cleanup/options.go [hunk 5]
```diff
@@ -24,8 +24,8 @@ type options struct {
 // Option - options for the cleanup chain element
 type Option func(*options)
 
-// WithCCClose - closes clientconn to prevent calling requests/closes on other datapath objects
-func WithCCClose() Option {
+// WithoutGrpcCall - closes client connection to prevent calling requests/closes on other endpoints
+func WithoutGrpcCall() Option {
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/cleanup/options.go:28]: Could you add `GRPC` into https://github.com/networkservicemesh/sdk/blob/main/staticcheck.conf and fix the linter issue?

Comment (glazychev-art) [reply pkg/networkservice/common/cleanup/options.go]: Sure! Done

**Review by denis-tingaikin — APPROVED** [review]

**Review by NikitaSkrynnik — APPROVED** [review]

**Pull request merged**